src/Stratis.Bitcoin/Base/InvalidBlockHashStore.cs

```diff
@@ -0,0 +1,154 @@
+        /// <param name="rejectedUntil">Time in UTC after which the block is no longer considered as invalid, or <c>null</c> if the block is to be considered invalid forever.</param>
+        void MarkInvalid(uint256 hashBlock, DateTime? rejectedUntil = null);
+    }
+
+    /// <summary>
+    /// In memory store of invalid block header hashes.
+    /// </summary>
+    /// <remarks>
+    /// The store has a limited capacity. When a new block header hash is marked as invalid 
+    /// once the capacity is reached, the oldest entry is removed and no longer considered invalid.
+    /// <para>
+    /// Entries with specified expiration time are either removed just like other entries - i.e. during 
+    /// an add operation when a capacity is reached, or they are removed when they are touched 
+    /// and it is detected that their expiration time is no longer in the future.
+    /// </para>
+    /// </remarks>
+    public class InvalidBlockHashStore : IInvalidBlockHashStore
+    {
+        /// <summary>Default value for the maximal number of hashes we can store.</summary>
+        public const int DefaultCapacity = 1000;
+
+        /// <summary>A provider of the date and time.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        /// <summary>Lock object to protect access to <see cref="invalidBlockHashes"/> and <see cref="orderedHashList"/>.</summary>
+        private readonly object lockObject = new object();
+
+        /// <summary>
+        /// Collection of blocks that are to be considered invalid. If the value of the entry is not <c>null</c>, 
+        /// the entry is considered invalid only for a certain amount of time.
+        /// </summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<uint256, DateTime?> invalidBlockHashes;
+
+        /// <summary>Circular array of block header hash entries to allow quick removal of the oldest entry once the capacity is reached.</summary>
+        /// <remarks>
+        /// All access to this object has to be protected by <see cref="lockObject"/>.
+        /// <para>The field is internal for testing purposes.</para>
+        /// </remarks>
+        internal readonly CircularArray<uint256> orderedHashList;
+
+        /// <summary>
+        /// Initializes the instance of the object.
+        /// </summary>
+        /// <param name="capacity">Maximal number of hashes we can store.</param>
+        public InvalidBlockHashStore(IDateTimeProvider dateTimeProvider, int capacity = DefaultCapacity)
+        {
+            this.dateTimeProvider = dateTimeProvider;
+
+            this.invalidBlockHashes = new Dictionary<uint256, DateTime?>(capacity);
+            this.orderedHashList = new CircularArray<uint256>(capacity);
+        }
+
+        /// <inheritdoc />
+        public bool IsInvalid(uint256 hashBlock)
+        {
+            bool res = false;
```

Comment:
This might be pedantic of me but can we name variables like this properly? I.e. "result", its just if we start shortening variables then whats stopping us from doing it everywhere? Lets keep naming consistent.

Reply:
we have a lot of `res` in other code base, thus seems fine
we also have a lot of "tx" and "txes" instead of "transaction(s)" 
i think this is consistent with what we have elsewhere

Reply:
That is exactly my point @Aprogiena ;) I don't think its good to continue that practice of shortening variables, so lets start fixing them going forward.

Reply:
ah i see, but i don't really agree :-) i quite like short names when it makes sense and does not change the meaning or is not confusing

so let me dodge the bullet by saying, i'd need further consensus to be happy with that

Reply:
personally i like class names to be little more explanatory, not that much for local variables

for example, i'm mostly perfectly happy with `i` in a (for) loop, only when clarity is better with something like `somethingIndex` i use it, but using `index` or `iterationIndex` every time i would not like

i there are many cases where short naming is perfectly fine and can't be confused - `tx` is always transaction and `res` is always result and `i` is always iteration index, etc. etc. i think there are many many cases where i'd rather save my wrists from typing more :-) and it won't hurt anything imho

Reply:
I tend to also prefer more explanatory names (even though our repo is full of `res`)
So to me its not a big deal.
I would whoever tend to prefer to use `index` over `i` come to think about it.

Reply:
@dangershony and @Aprogiena can we then decide on what to use throughout? We can't have variable naming short in some areas and explanatory in other areas? I dont mind either way, just want consistency 👍

Reply:
i'd be happy with a rule such as `variable/class name must not be confusing, or too similar to some other variable/class that is unrelated (so singular and plural variables are OK like foreach (item in items), the name must be explanatory. and then if the variable/class name is as such, it should be as short as possible`

with such a rule, in almost all code (unless there are conflicts with something else) `res` is better than `result`, `tx` is better than `transaction`, `id` is better than `identifier`, `txid` is better than `transactionIdentifer`

Reply:
@Aprogiena omg this discussion is getting out of hand. It would've been so much easier if you just renamed the variable. Then at least we have some consistency. Cause now, with your way, we need to define what variables can be shortened or not. (And I'm not referring to "for" or while variables etc). @dangershony

Reply:
@fassadlr  i understand what you mean, but i just don't agree with you :-) so i'm trying to avoid such a change by thinking about it. and so far as i've thought about it i find that i'm ok with my original position, so i still don't want the change :-)

Reply:
there are many things that i just do by intuition and have never really supported those things with any rational arguments that would make sense to me. so it makes sense to me, if something is a source of controversy, to sort my thoughts about it and either change my opinion or not. here i still consider my naming as rational and better than the proposed alternative